explorer/interpreter/type_checker.cpp

```diff
@@ -2920,6 +2920,354 @@ auto TypeChecker::TypeCheckPattern(
+      }
+    } else if (auto* value = value_.dyn_cast<const Value*>()) {
+      if (auto* alt = dyn_cast<AlternativeValue>(value)) {
+        return alt->alt_name();
+      }
+    }
+    return {};
+  }
+
+  auto NumElements() const -> int {
+    if (auto* pattern = value_.dyn_cast<const Pattern*>()) {
+      if (auto* tuple_pattern = dyn_cast<TuplePattern>(pattern)) {
+        return tuple_pattern->fields().size();
+      } else if (isa<AlternativePattern>(pattern)) {
+        return 1;
+      }
+    } else if (auto* value = value_.dyn_cast<const Value*>()) {
+      if (auto* tuple = dyn_cast<TupleValue>(value)) {
+        return tuple->elements().size();
+      } else if (auto* alt = dyn_cast<AlternativeValue>(value)) {
+        return 1;
+      }
+    }
+    return 0;
+  }
+
+  void AppendElementsTo(std::vector<AbstractPattern>& out) const {
+    if (auto* pattern = value_.dyn_cast<const Pattern*>()) {
+      if (auto* tuple_pattern = dyn_cast<TuplePattern>(pattern)) {
+        auto fields = tuple_pattern->fields();
+        out.insert(out.end(), fields.begin(), fields.end());
+      } else if (auto* alt_pattern = dyn_cast<AlternativePattern>(pattern)) {
+        out.push_back(&alt_pattern->arguments());
+      }
+    } else if (auto* value = value_.dyn_cast<const Value*>()) {
+      if (auto* tuple = dyn_cast<TupleValue>(value)) {
+        auto* tuple_type = cast<TupleValue>(type_);
+        assert(tuple->elements().size() == tuple_type->elements().size());
+        for (size_t i = 0; i != tuple->elements().size(); ++i) {
+          out.push_back(
+              AbstractPattern(tuple->elements()[i], tuple_type->elements()[i]));
+        }
+      } else if (auto* alt = dyn_cast<AlternativeValue>(value)) {
+        out.push_back(AbstractPattern(
+            &alt->argument(),
+            *cast<ChoiceType>(type_)->FindAlternative(alt->alt_name())));
+      }
+    }
+  }
+
+  auto value() const -> const Value& { return *value_.get<const Value*>(); }
+
+  auto type() const -> const Value& {
+    assert(kind() == Compound);
+    return *type_;
+  }
+
+ private:
+  struct alignas(8) WildcardTag {};
+  AbstractPattern(WildcardTag)
+      : value_(static_cast<const WildcardTag*>(nullptr)), type_(nullptr) {}
+
+  void Set(Nonnull<const Pattern*> pattern) {
+    type_ = &pattern->static_type();
+    switch (pattern->kind()) {
+      case PatternKind::AddrPattern:
+      case PatternKind::AutoPattern:
+      case PatternKind::BindingPattern:
+      case PatternKind::GenericBinding:
+        value_ = static_cast<const WildcardTag*>(nullptr);
+        break;
+
+      case PatternKind::TuplePattern:
+      case PatternKind::AlternativePattern:
+        value_ = pattern;
+        break;
+
+      case PatternKind::ExpressionPattern:
+        value_ = &pattern->value();
+        break;
+
+      case PatternKind::VarPattern:
+        Set(&cast<VarPattern>(pattern)->pattern());
+        break;
+    }
+  }
+
+ private:
```

Comment:
Multiple private sections?

Reply:
Yeah, this is an ingrained habit from a non-Google coding style to have different `private` sections for member functions versus data. I'll try to stop :)

```suggestion
```